docs/guides/implementation-guides/how-to-implement-dependency-injection-with-pure-di.md

```diff
@@ -0,0 +1,207 @@
+{
+    public string CreateGreetings()
+    {
+        repository.RegisterSomething();
+        return "Example of Dependency Injection implementation using Pure.DI";
+    }
+}
+
+public interface IRepository
+{
+    void RegisterSomething();
+}
+
+public class Repository: IRepository
+{
+    public void RegisterSomething()
+    {
+    }
+}
+```
+
+## Step 1: Install the NuGet package for DI
+
+Run the following command in a terminal inside your project directory to install the [Pure.DI](https://www.nuget.org/packages/Pure.DI) source generator package.
+
+```shell
+dotnet add package Pure.DI
+```
+
+## Step 2: Add a composition class
+
+This class setups how the composition of objects will be created for the application.
+
+```csharp
+using Pure.DI;
+using static Pure.DI.Lifetime;
+
+public partial class Composition
+{
+    private static void Setup() => DI.Setup()
+        .Root<MainWindow>(nameof(MainWindow))
+        .Root<IMainViewModel>(nameof(MainViewModel))
+
+        .Bind().As(Singleton).To<Repository>()
+        .Bind().To<BusinessService>()
+        .Bind().As(Singleton).To<MainViewModel>();
+}
+```
+
+Advantages over classical DI container libraries:
+- No performance impact or side effects when creating composition of objects.
+- All logic for analyzing the graph of objects, constructors and methods takes place at compile time. Pure.DI notifies the developer at compile time of missing or cyclic dependencies, cases when some dependencies are not suitable for injection, etc.
+- Does not add dependencies to any additional assembly.
+- Since the generated code uses primitive language constructs to create object compositions and does not use any libraries, you can easily debug the object composition code as regular code in your application.
+
+## Step 3: Modify App.axaml.cs
+
+Next, the `App.xaml.cs` class should be modified to use the main window created in pure DI paradigm: 
+
+```csharp
+public class App : Application
+{
+    public override void Initialize()
+    {
+        AvaloniaXamlLoader.Load(this);
+    }
+
+    public override void OnFrameworkInitializationCompleted()
+    {
+        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop
+            && Resources[nameof(Composition)] is Composition composition)
+        {
+            desktop.MainWindow = composition.MainWindow;
+        }
+        
+        base.OnFrameworkInitializationCompleted();
+    }
+}
+```
+
+This code is only needed to create a main window in a pure DI paradigm using a shared composition instance:
+
+```csharp
+if (Resources[nameof(Composition)] is Composition composition)
+{
+    desktop.MainWindow = composition.MainWindow;
+}
+```
+
+This will be useful if the main window will require dependencies to be injected.
+
+Advantages over classical DI container libraries:
+- No explicit initialisation of data contexts is required. Data contexts are configured directly in `.axaml` files according to the MVVM approach.
```

Comment:
Pure.DI doesn't create a container like in classic DI container frameworks. Pure.DI is a source code generator. For the example in this document, it generates a class like this:

```c#
AvaloniaSimpleApp namespace;

partial class Composition
{
    private readonly Lock _lock = new();

    private MainViewModel? _mainViewModel;
    private Repository? _repository;

    public IMainViewModel MainViewModel
    {
        get
        {
            if (_mainViewModel is null)
            {
                using (_lock.EnterScope())
                {
                    if (_mainViewModel is null)
                    {
                        if (_repository is null)
                        {
                            _repository = new Repository();
                        }

                        _mainViewModel = new MainViewModel(new BusinessService(_repository));
                    }
                }
            }

            return _mainViewModel;
        }
    }

    public MainWindow MainWindow => new()
}
```

This class is roughly what you would write if you were doing object composition in a pure DI paradigm by hand. There aren't any dependencies on third party libraries, you don't need methods like `GetService(Type type type)`. You have a class property for each root of the composition.

How many more steps are you talking about? All you need to do is to define one instance of the class in the resources to create the roots of the composition, and that's it. Except for the settings themselves for defining the dependency graph, there is no more code, everything is customized by bindings inside xaml files. Just count the number of lines of code in the examples. 

> There is also really nothing stopping you from setting up MS DI in a similar way. It just requires slightly more effort (and also a static locator).

I suggest you try doing this for MS DI, I'm not sure it will look as easy, at least not because any classic DI library has a Service Locator at the very beginning. I'm not claiming that the MS DI approach somehow violates MVVM, I'm just pointing out that the presentation layer in the Pure.DI approach is even more clearly separated from the other layers and is consistent with MVVM. The presentation layer has no .cs files and no tricks to initialize data contexts.